pgml-cms/blog/semantic-search-in-postgres-in-15-minutes.md

```diff
@@ -0,0 +1,467 @@
+| 2 | I think tomatoes are incredible on burgers. |
+
+
+and a user is looking for the answer to the question: "What is the pgml.transform function?". If we embed the search query and all of the documents using a model like _mixedbread-ai/mxbai-embed-large-v1_, we can compare the query embedding to all of the document embeddings, and select the document that has the closest embedding in vector space, and therefore in meaning, to the to the answer. 
+
+These are big embeddings, so we can’t simply estimate which one is closest. So, how do we actually measure the similarity (distance) between different vectors?
+
+_pgvector_ as of this writing supports four different measurements of vector similarity:
+
+- L2 distance
+- (negative) inner product
+- cosine distance
+- L1 distance
+
+For most use cases we recommend using the cosine distance as defined by the formula:
+
+<figure><img src=".gitbook/assets/cosine_similarity.png" alt="cosine similarity formula"><figcaption></figcaption></figure>
+
+where A and B are two vectors. 
+
+This is a somewhat confusing formula but luckily  _pgvector_ provides an operator that computes the cosine distance for us:
+
+!!! generic
+
+!!! code_block time="64.643 ms"
+
+```postgresql
+SELECT '[1,2,3]'::vector <=> '[2,3,4]'::vector;
+```
+
+!!!
+
+!!! results
+
+```text
+   cosine_distance    
+----------------------
+ 0.007416666029069763
+```
+
+!!!
+
+!!!
+
+Other distance functions have similar formulas and provide convenient operators to use as well. It may be worth testing other operators and to see which performs better for your use case. For more information on the other distance functions, take a look at our [Embeddings guide](https://postgresml.org/docs/guides/embeddings/vector-similarity).
+
+Going back to our search example, we can compute the cosine distance between our query embedding and our documents:
+
+!!! generic
+
+!!! code_block
+
+```postgresql
+SELECT pgml.embed(
+  'mixedbread-ai/mxbai-embed-large-v1',
+  'What is the pgml.transform function?'
+)::vector
+  <=>
+pgml.embed(
+  'mixedbread-ai/mxbai-embed-large-v1',
+  'The pgml.transform function is a PostgreSQL function for calling LLMs in the database.'
+)::vector AS cosine_distance;
+
+SELECT pgml.embed(
+  'mixedbread-ai/mxbai-embed-large-v1',
+  'What is the pgml.transform function?'
+)::vector
+  <=>
+pgml.embed(
+  'mixedbread-ai/mxbai-embed-large-v1',
+  'I think tomatoes are incredible on burgers.'
+)::vector AS cosine_distance;
+```
+
+!!!
+
+!!! results
+
+```text
+cosine_distance   
+--------------------
+ 0.1114425936213167
+
+cosine_distance   
+--------------------
+ 0.7328613577628744
+```
+
+!!!
+
+!!!
+
+You'll notice that the distance between "What is the pgml.transform function?" and "The pgml.transform function is a PostgreSQL function for calling LLMs in the database." is much smaller than the cosine distance between "What is the pgml.transform function?" and "I think tomatoes are incredible on burgers".
```

Comment:
It's probably worth mentioning "out of sample" tokens, since they are rife on domain specific topics like this.

Reply:
I'm not sure what that is?